### PR TITLE
Reduce test_scale_out delay

### LINKED ISSUE #1885: Possible race condition in htex scale down
**Describe the bug**
I've seen the below test failure as a one-off in `--config local` tests in the CI on several occasions.

My gut feeling is there's a race condition here - either in the test, or in parsl itself. I'm noting this for later investigation.

The scale down behaviour here is relatively new. Tagging @yadudoc who implemented it.

```
parsl/tests/test_scaling/test_scale_down.py F
=================================== FAILURES ===================================
________________________________ test_scale_out ________________________________
    @pytest.mark.local
    def test_scale_out():
        dfk = parsl.dfk()
    
        # Since we have init_blocks = 0, at this point we should have 0 managers
        print("Executor : ", dfk.executors['htex_local'])
        print("Before")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
        assert len(dfk.executors['htex_local'].connected_managers) == 0, "Expected 0 managers at start"
    
        fus = [sleeper(i) for i in [3, 3, 25, 25, 50]]
    
        for i in range(2):
            fus[i].result()
    
        # At this point, since we have 1 task already processed we should have atleast 1 manager
        print("Between")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
>       assert len(dfk.executors['htex_local'].connected_managers) == 5, "Expected 5 managers once tasks are running"
E       AssertionError: Expected 5 managers once tasks are running
E       assert 3 == 5
E        +  where 3 = len([{'active': True, 'block_id': '1', 'idle_duration': 0, 'manager': '884d27969893', ...}, {'active': True, 'block_id': '...486f', ...}, {'active': True, 'block_id': '2', 'idle_duration': 0.0022573471069335938, 'manager': 'a19000850efd', ...}])
E        +    where [{'active': True, 'block_id': '1', 'idle_duration': 0, 'manager': '884d27969893', ...}, {'active': True, 'block_id': '...486f', ...}, {'active': True, 'block_id': '2', 'idle_duration': 0.0022573471069335938, 'manager': 'a19000850efd', ...}] = HighThroughputExecutor(\n    address=None, \n    address_probe_timeout=None, \n    cores_per_worker=1.0, \n    cpu_affinit...e, \n    worker_logdir_root=None, \n    worker_port_range=(54000, 55000), \n    worker_ports=None, \n    working_dir=None\n).connected_managers
parsl/tests/test_scaling/test_scale_down.py:66: AssertionError
```

**To Reproduce**
unlikely - this looks like a low probability race condition to me.

**Expected behavior**
test should pass

**Environment**
CI, PR #1884

### PR DESCRIPTION
**Pull request by khk-globus**
Locally, I was consistently seeing this test take 56s.  This commit reduces that to 12s locally.

There are two main changes:

- Bind to 127.0.0.1, rather than to all live interfaces.  Some of the delay was from the network stack

- Replace the littered `time.sleep()` calls with an "Event" based approach (polled files on the filesystem)

The other addition of note is `try_assert()`.  This my approximation of JUnit's `assertTimeout` or QT's `QTRY_VERIFY`: continually try a test condition for a timeout period.

I'm not able to discern at this time if this robustly resolves #1885.

## Type of change

- Code maintenance/cleanup

### CODE REVIEW

File: parsl/tests/test_scaling/test_scale_down.py [hunk 1]
```diff
@@ -1,80 +1,97 @@
...
+    futs = [waiting_app(i, inputs=inputs) for i in range(ntasks)]
 
-    logger.info("asserting 5 managers")
-    assert len(dfk.executors['htex_local'].connected_managers()) == 5, "Expected 5 managers after some time"
+    while ready_path.read_text().count("\n") < _max_blocks:
+        time.sleep(0.5)
 
-    logger.info("waiting for all futures to complete")
-    [x.result() for x in fus]
+    assert len(dfk.executors['htex_local'].connected_managers()) == 5
```
Comment (benclifford) [inline parsl/tests/test_scaling/test_scale_down.py:83]: that explicit 5 should probably be _max_blocks

File: parsl/tests/test_scaling/test_scale_down.py [hunk 2]
```diff
@@ -1,80 +1,97 @@
...
-    time.sleep(25)
+    assert dfk.executors['htex_local'].outstanding == 0
 
-    logger.info("asserting 2 managers remain")
-    assert len(dfk.executors['htex_local'].connected_managers()) == 2, "Expected 2 managers when no tasks, lower bound by min_blocks"
+    def assert_kernel():
+        return len(dfk.executors['htex_local'].connected_managers()) == _min_blocks
 
-    logger.info("test passed")
+    try_assert(
```
Comment (benclifford) [inline parsl/tests/test_scaling/test_scale_down.py:93]: some part of me says `try_assert` adds a lot of complexity that might be a relatively small while loop?

File: parsl/tests/test_scaling/test_scale_down.py [hunk 3]
```diff
@@ -1,80 +1,97 @@
...
+        assert_kernel,
+        fail_msg="Expected 2 managers when no tasks, lower bound by min_blocks",
```
Comment (benclifford) [inline parsl/tests/test_scaling/test_scale_down.py:95]: 2 should not be explicit

File: parsl/tests/test_scaling/test_scale_down.py (hunk 1 above)
Comment (khk-globus) [reply parsl/tests/test_scaling/test_scale_down.py:83]: Bah, of course.  Got halfway there ...

File: parsl/tests/test_scaling/test_scale_down.py (hunk 2 above)
Comment (khk-globus) [reply parsl/tests/test_scaling/test_scale_down.py:93]: Yeah, I concur on this point.  However, my thinking was that surely this won't be the only place where we need a "give it some settling time" type assertion.

As discussed out-of-band, neither of us has a particular strong global insight into all of the tests, so agreed to leave it around for now.  If it ends up being used only by this test, we can remove it at some point in the indeterminate future.

File: parsl/tests/conftest.py [hunk 4]
```diff
@@ -290,3 +294,67 @@ def pytest_ignore_collect(path):
...
+                time.sleep(check_period_s)
+            else:
+                att_fail = (
+                    f"\n  (Still failing after attempt limit [{attempts}], testing"
+                    f" every {check_period_ms}ms)"
+                )
+                exc = AssertionError(f"{str(fail_msg)}{att_fail}".strip())
+                raise exc.with_traceback(tb)
+
+        elif timeout_s > 0:
```
Comment (benclifford) [inline parsl/tests/conftest.py:344]: this if/elif structure means that timeout_ms will be ignored if attempts !=0, right?

Comment (khk-globus) [reply parsl/tests/conftest.py:344]: That is correct; an intentional choice.

**Pull request merged**